### ISSUE TITLE
qemu job failed at "rpc error: code = Unknown desc = unable to configur e cgroups: no such file or directory"

### ISSUE THREAD

**Issue by joechchen** [hashicorp/nomad#23250]
### Nomad version

Nomad v1.8.0
BuildDate 2024-05-28T17:38:17Z
Revision 28b82e4b2259fae5a62e2ed47395334bea5a24c4

### Operating system and Environment details
Ubuntu 22.04.4 LTS
### Issue

qemu job failed at:
```
...
2024-06-06T13:45:20.490-0400 [ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 task=freedos error="rpc error: code = Unknown desc = unable to configur
e cgroups: no such file or directory"
...
```
The same job runs fine on v1.7.1 and prior.

### Reproduction steps
```
nomad run freedos.nomad
nomad monitor -log-level=DEBUG
```
#### Expected Result

The job should run.

#### Actual Result

The job failed.

### Job file (if appropriate)
```
job "freedos" {
  datacenters = ["dc1"]

  group "g1" {
    task "freedos" {
      resources {
        cpu    = 500
        memory = 256
      }

      artifact {
        source      = "https://github.com/angrycub/nomad_example_jobs/raw/main/applications/vms/freedos/freedos.img.tgz"
        destination = "local"
        options {
          checksum  = "sha256:8d2817126bf46ba2b4fca0b0c49eed2cc208c6f6448651e82c6d973fcba36569"
        }
      }

      driver = "qemu"

      config {
        image_path  = "local/freedos.img"
        accelerator = "kvm"
        args = []
      }
    }
  }
}
```

### Nomad Client logs (if appropriate)
```
2024-06-06T13:45:20.436-0400 [DEBUG] client.artifact: sub-process: OUTPUT="artifact download was a success"
2024-06-06T13:45:20.436-0400 [TRACE] client.alloc_runner.task_runner.task_hook.procisolation: setting up client process management: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 task=freedos task=40bd6120/freed
os
2024-06-06T13:45:20.437-0400 [TRACE] client.proclib: setup client process management: task=40bd6120/freedos
2024-06-06T13:45:20.437-0400 [TRACE] client.proclib: initialize cgroup: task=40bd6120/freedos
2024-06-06T13:45:20.438-0400 [DEBUG] client.driver_mgr.qemu: starting QEMU VM command : driver=qemu args="/usr/bin/qemu-system-x86_64 -machine type=pc,accel=kvm -name freedos.img -m 256M -drive file=local/freed
os.img,if=ide -nographic -vnc 127.0.0.1:1 -enable-kvm -cpu host"
2024-06-06T13:45:20.438-0400 [DEBUG] client.driver_mgr.qemu.executor: starting plugin: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 driver=qemu task_name=freedos path=/usr/bin/nomad args=["/usr/bin/nomad", "ex
ecutor", "{\"LogFile\":\"/opt/nomad/data/alloc/40bd6120-39b6-5f19-7416-322de0c20ad5/freedos/freedos-executor.out\",\"LogLevel\":\"debug\",\"FSIsolation\":false,\"Compute\":{\"tc\":15600,\"nc\":4}}"]
2024-06-06T13:45:20.439-0400 [DEBUG] client.driver_mgr.qemu.executor: plugin started: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 driver=qemu task_name=freedos path=/usr/bin/nomad pid=2667699
2024-06-06T13:45:20.439-0400 [DEBUG] client.driver_mgr.qemu.executor: waiting for RPC address: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 driver=qemu task_name=freedos plugin=/usr/bin/nomad
2024-06-06T13:45:20.485-0400 [DEBUG] client.driver_mgr.qemu.executor: using plugin: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 driver=qemu task_name=freedos version=2
2024-06-06T13:45:20.486-0400 [TRACE] client.driver_mgr.qemu.executor.stdio: waiting for stdio data: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 driver=qemu task_name=freedos
2024-06-06T13:45:20.488-0400 [DEBUG] client.driver_mgr.qemu.executor.stdio: received EOF, stopping recv loop: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 driver=qemu task_name=freedos err="rpc error: code = U
navailable desc = error reading from server: EOF"
2024-06-06T13:45:20.490-0400 [INFO]  client.driver_mgr.qemu.executor: plugin process exited: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 driver=qemu task_name=freedos plugin=/usr/bin/nomad id=2667699
2024-06-06T13:45:20.490-0400 [DEBUG] client.driver_mgr.qemu.executor: plugin exited: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 driver=qemu task_name=freedos
2024-06-06T13:45:20.490-0400 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 task=freedos type="Driver Failure" msg="rpc error: code = Unknown desc = unable to
 configure cgroups: no such file or directory" failed=false
2024-06-06T13:45:20.490-0400 [ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 task=freedos error="rpc error: code = Unknown desc = unable to configur
e cgroups: no such file or directory"
2024-06-06T13:45:20.490-0400 [INFO]  client.alloc_runner.task_runner: not restarting task: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 task=freedos reason="Error was unrecoverable"
2024-06-06T13:45:20.490-0400 [TRACE] client.alloc_runner.task_runner: setting task state: alloc_id=40bd6120-39b6-5f19-7416-322de0c20ad5 task=freedos state=dead
```

**Comment by fjoenichols** [comment]
I am also seeing the same behavior on nomad v1.8.0 attempting to use the qemu example at https://github.com/angrycub/nomad_example_jobs/blob/main/qemu/tc_ssh.nomad

```
job "j1" {
  datacenters = ["dc1"]

  group "g1" {

    network {
      port "http" { 
        to = -1
      }
      port "ssh" {
        to = -1
      }
    }

    service {
      tags = ["tag1"]
      port = "http"

      check {
        type     = "http"
        port     = "http"
        path     = "/index.html"
        interval = "10s"
        timeout  = "2s"
      }
    }

    task "t1" {
      template {
        data = <<EOH
      Guest System
      EOH

        destination = "local/index.html"
      }

      artifact {
        source = "https://github.com/angrycub/nomad_example_jobs/raw/main/qemu/tinycore.qcow2"
        destination = "local/images"
      }

      driver = "qemu"

      config {
        image_path = "local/images/tinycore.qcow2"

        ## Uncomment if KVM is available on your system
        accelerator = "kvm"

        args = [
          "-device",
          "e1000,netdev=user.0",
          "-netdev",
          "user,id=user.0,hostfwd=tcp::${NOMAD_PORT_http}-:80,hostfwd=tcp::${NOMAD_PORT_ssh}-:22",
          "-drive", "file=fat:rw:./local,format=raw,media=disk"
        ]
      }
    }
  }
}
```

**Comment by tgross** [comment]
Hi @joechchen! It looks like the task driver is trying to configure cgroups for the process we use to manage QEMU, but that's failing for some reason related to the cgroup setup on the host.

A couple of questions:
* Are you running the Nomad client as root?
* Are you running Nomad in a custom cgroup?
* Can you provide the logs for the `executor` process? Those can be found in the allocation directory after a failure, in a location like: `/opt/nomad/alloc/5c5b53f6-cee4-e164-760f-fdfbdf46683f/t1/t1-executor.out` (this is from your client logs).

**Comment by mwright-pivotal** [comment]
hello @tgross - I am seeing this in my homelab as well.  Nomad on Ubuntu 22.04 bare metal and KVM support

I see this log output in the alloc directory:

root@host2:/opt/nomad/client/client_datadir/alloc/24e03b91-5d09-d7bc-4149-71eafdc4fa89/virtual# ls
local  private  secrets  tmp  virtual-executor.out
root@host2:/opt/nomad/client/client_datadir/alloc/24e03b91-5d09-d7bc-4149-71eafdc4fa89/virtual# cat virtual-executor.out
{"@level":"debug","@message":"plugin address","@timestamp":"2024-06-25T20:03:45.927311Z","address":"/tmp/plugin2125446520","network":"unix"}
{"@level":"error","@message":"failed to configure container, process isolation will not work","@module":"executor","@timestamp":"2024-06-25T20:03:45.930127Z","error":"no such file or directory"}

my job definition is here: https://github.com/mwright-pivotal/learn-terraform-cloud-agents/blob/main/windows2022vm-job.hcl

**Comment by mwright-pivotal** [comment]
I am running the client as root but have not done anything to setup a custom cgroup

**Comment by tgross** [comment]
Ok, thanks folks. I did what I should have done a few days ago when I first touched this issue, which is to run the example job @fjoenichols mentioned here: https://github.com/hashicorp/nomad/issues/23250#issuecomment-2158591791. I can reproduce the problem exactly now in my local development environment on Linux. I'll dig into this and report back when I know more.

**Comment by tgross** [comment]
Draft PR is up here: https://github.com/hashicorp/nomad/pull/23466

**Comment by tgross** [comment]
#23466 has been merged and will ship in Nomad 1.8.2 (with backports to Nomad Enterprise 1.7.x and 1.6.x )

**Comment by djthorpe** [comment]
Sorry - I created a new issue here:
https://github.com/hashicorp/nomad/issues/23595
